File: JavaToKotlin/app/src/main/java/io/dustin/apps/common/code/BoardType.java
package io.dustin.apps.common.code;

import io.dustin.apps.board.domain.like.model.LikeCountService;
import io.dustin.apps.common.utils.CommonUtil;
import io.dustin.apps.common.utils.Supplier;

public enum BoardType {

    POSTING("posting", () -> CommonUtil.getBean("posting", LikeCountService.class)),
    COMMENT("comment", () -> CommonUtil.getBean("comment", LikeCountService.class)),
    NOTICE("notice", () -> CommonUtil.getBean("notice", LikeCountService.class));



    private String code;

    private Supplier<LikeCountService> supplier;

    BoardType(String code, Supplier<LikeCountService> supplier) {
        this.code = code;
        this.supplier = supplier;
    }

    public LikeCountService getBean() {
        return this.supplier.get();
    }

}

File: JavaToKotlin/apps(kotlin)/src/main/kotlin/io/dustin/apps/common/code/BoardType.kt
package io.dustin.apps.common.code

import io.dustin.apps.board.domain.like.model.LikeCountService

enum class BoardType(private val code: String?, supplier: Supplier<LikeCountService?>?) {
    POSTING("posting", Supplier<LikeCountService?> { CommonUtil.getBean("posting", LikeCountService::class.java) }),
    COMMENT("comment", Supplier<LikeCountService?> { CommonUtil.getBean("comment", LikeCountService::class.java) }),
    NOTICE("notice", Supplier<LikeCountService?> { CommonUtil.getBean("notice", LikeCountService::class.java) });

    private val supplier: Supplier<LikeCountService?>?

    init {
        this.supplier = supplier
    }

    fun getBean(): LikeCountService? {
        return supplier.get()
    }
}
